### PR TITLE
avoid access to exprt::opX()

### PR DESCRIPTION
**Pull request by kroening**
This improves memory safety.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a7b7c28a724b025a787bfca6b69ce074134b9334).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114577110

File: src/analyses/goto_check.cpp [hunk 1]
```diff
@@ -627,16 +621,13 @@ void goto_checkt::integer_overflow_check(
...
   if(expr.id()==ID_div)
   {
-    assert(expr.operands().size()==2);
-
     // undefined for signed division INT_MIN/-1
     if(type.id()==ID_signedbv)
     {
       equal_exprt int_min_eq(
-        expr.op0(), to_signedbv_type(type).smallest_expr());
+        to_div_expr(expr).op0(), to_signedbv_type(type).smallest_expr());
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: op0 -> dividend

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -627,16 +621,13 @@ void goto_checkt::integer_overflow_check(
...
 
-      equal_exprt minus_one_eq(
-        expr.op1(), from_integer(-1, type));
+      equal_exprt minus_one_eq(to_div_expr(expr).op1(), from_integer(-1, type));
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: op1 -> divisor

File: src/analyses/goto_check.cpp [hunk 3]
```diff
@@ -880,11 +870,9 @@ void goto_checkt::float_overflow_check(
   }
   else if(expr.id()==ID_div)
   {
-    assert(expr.operands().size()==2);
-
     // Can overflow if dividing by something small
     const isinf_exprt new_inf(expr);
-    const isinf_exprt op0_inf(expr.op0());
+    const isinf_exprt op0_inf(to_div_expr(expr).op0());
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: dividend

File: src/analyses/goto_check.cpp [hunk 4]
```diff
@@ -915,8 +903,8 @@ void goto_checkt::float_overflow_check(
     {
       // Can overflow
       const isinf_exprt new_inf(expr);
-      const isinf_exprt op0_inf(expr.op0());
-      const isinf_exprt op1_inf(expr.op1());
+      const isinf_exprt op0_inf(to_binary_expr(expr).op0());
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: lhs

File: src/analyses/goto_check.cpp [hunk 5]
```diff
@@ -915,8 +903,8 @@ void goto_checkt::float_overflow_check(
...
+      const isinf_exprt op1_inf(to_binary_expr(expr).op1());
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: rhs

File: src/analyses/goto_check.cpp [hunk 6]
```diff
@@ -970,16 +958,18 @@ void goto_checkt::nan_check(
...
+    const auto &div_expr = to_div_expr(expr);
 
     // there a two ways to get a new NaN on division:
     // 0/0 = NaN and x/inf = NaN
     // (note that x/0 = +-inf for x!=0 and x!=inf)
     const and_exprt zero_div_zero(
-      ieee_float_equal_exprt(expr.op0(), from_integer(0, expr.op0().type())),
-      ieee_float_equal_exprt(expr.op1(), from_integer(0, expr.op1().type())));
+      ieee_float_equal_exprt(
+        div_expr.op0(), from_integer(0, div_expr.op0().type())),
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: op0/1 -> dividend/sor

File: src/analyses/goto_check.cpp [hunk 7]
```diff
@@ -988,16 +978,18 @@ void goto_checkt::nan_check(
...
       return nan_check(make_binary(expr), guard);
 
-    assert(expr.operands().size()==2);
+    const auto &mult_expr = to_mult_expr(expr);
 
     // Inf * 0 is NaN
     const and_exprt inf_times_zero(
-      isinf_exprt(expr.op0()),
-      ieee_float_equal_exprt(expr.op1(), from_integer(0, expr.op1().type())));
+      isinf_exprt(mult_expr.op0()),
```
Comment (smowton) [inline src/analyses/goto_check.cpp:978]: op0/1 -> lhs/rhs

File: src/analyses/goto_check.cpp [hunk 8]
```diff
@@ -1006,41 +998,41 @@ void goto_checkt::nan_check(
...
-      or_exprt(
-        and_exprt(
-          equal_exprt(expr.op0(), minus_inf),
-          equal_exprt(expr.op1(), plus_inf)),
-        and_exprt(
-          equal_exprt(expr.op0(), plus_inf),
-          equal_exprt(expr.op1(), minus_inf)));
+    isnan = or_exprt(
+      and_exprt(
+        equal_exprt(plus_expr.op0(), minus_inf),
```
Comment (smowton) [inline src/analyses/goto_check.cpp:1004]: op0/1 -> lhs/rhs

File: src/analyses/goto_check.cpp [hunk 9]
```diff
@@ -1006,41 +998,41 @@ void goto_checkt::nan_check(
...
-      or_exprt(
-        and_exprt(
-          equal_exprt(expr.op0(), plus_inf),
-          equal_exprt(expr.op1(), plus_inf)),
-        and_exprt(
-          equal_exprt(expr.op0(), minus_inf),
-          equal_exprt(expr.op1(), minus_inf)));
+    isnan = or_exprt(
+      and_exprt(
+        equal_exprt(minus_expr.op0(), plus_inf),
```
Comment (smowton) [inline src/analyses/goto_check.cpp:1024]: op0/1- > lhs/rhs

File: src/analyses/invariant_set.cpp [hunk 10]
```diff
@@ -506,7 +507,7 @@ void invariant_sett::strengthen_rec(const exprt &expr)
 
       forall_operands(it, bitand_op)
       {
-        exprt tmp(equal_expr);
```
Comment (smowton) [inline src/analyses/invariant_set.cpp:509]: why `auto`?

File: src/analyses/invariant_set.cpp [hunk 11]
```diff
@@ -516,7 +517,7 @@ void invariant_sett::strengthen_rec(const exprt &expr)
     }
     else if(equal_expr.op0().id() == ID_bitand)
     {
-      exprt tmp(equal_expr);
+      auto tmp(equal_expr);
```
Comment (smowton) [inline src/analyses/invariant_set.cpp:520]: why `auto`?

File: src/analyses/invariant_set.cpp [hunk 12]
```diff
@@ -757,7 +758,8 @@ void invariant_sett::nnf(exprt &expr, bool negate)
     {
       // !a<=b <-> !b=>a <-> b<a
       expr.id(ID_lt);
-      std::swap(expr.op0(), expr.op1());
+      auto &rel = to_binary_relation_expr(expr);
+      std::swap(rel.op0(), rel.op1());
```
Comment (smowton) [inline src/analyses/invariant_set.cpp]: rhs/lhs (here and a few below)

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (kroening) [reply src/analyses/goto_check.cpp]: done!

File: src/analyses/goto_check.cpp (hunk 4 above)
Comment (kroening) [reply src/analyses/goto_check.cpp]: That method probably should exist, but doesn't.

File: src/analyses/invariant_set.cpp (hunk 10 above)
Comment (kroening) [reply src/analyses/invariant_set.cpp:509]: That means that the type of `tmp` will allow us to access .opX().

File: src/analyses/invariant_set.cpp (hunk 12 above)
Comment (kroening) [reply src/analyses/invariant_set.cpp]: done!

File: src/analyses/goto_check.cpp [hunk 13]
```diff
@@ -844,15 +837,14 @@ void goto_checkt::float_overflow_check(
...
+    const auto &op = to_typecast_expr(expr).op();
+    if(op.type().id() == ID_floatbv)
     {
       // float-to-float
-      const isinf_exprt op0_inf(expr.op0());
+      const isinf_exprt op_inf(op);
       const isinf_exprt new_inf(expr);
 
-      or_exprt overflow_check(op0_inf, not_exprt(new_inf));
+      or_exprt overflow_check(op_inf, not_exprt(new_inf));
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Nit picking: Do names like "op_inf" and "new_inf" really carry sufficient value to justify those temporaries? If they do, we could actually `std::move` them at least.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a9a0a01e853c61725dfebdd15ecdab1bd9a1bd28).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116179208
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 33705a254f4b286706fbd03044f7c73effa83d01).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116183097

File: src/analyses/goto_check.cpp (hunk 13 above)
Comment (kroening) [reply src/analyses/goto_check.cpp]: ok, inlined

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fd32f5bb2b889bd49d37fdf2daf1c7e61dfac176).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116499820

**Review by owen-mc-diffblue — APPROVED** [review]

**Pull request merged**